Subject: [PATCH] git-remote-mediawiki: get correct maximum revision id

The get_last_global_remote_rev function attempts to determine the largest
revision number by examining the most recent change from the recent changes
list.  However, the "revid" value may be "0" in some cases when the change
is not located in the 'Main' (NS_MAIN = 0) name space.

Compensate for this by making one additional query when the "revid" value
comes back as "0" in order to get the real "revid" value for the change
in question thereby correctly determining the largest revision number in
those cases.

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>

---
 contrib/mw-to-git/git-remote-mediawiki.perl | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/contrib/mw-to-git/git-remote-mediawiki.perl b/contrib/mw-to-git/git-remote-mediawiki.perl
index 89cb879f..cbba8f39 100755
--- a/contrib/mw-to-git/git-remote-mediawiki.perl
+++ b/contrib/mw-to-git/git-remote-mediawiki.perl
@@ -508,7 +508,24 @@ sub get_last_global_remote_rev {
 		rcdir => 'older',
 	};
 	my $result = $mediawiki->api($query);
-	return $result->{query}->{recentchanges}[0]->{revid};
+	my $revid = $result->{query}->{recentchanges}[0]->{revid};
+	if (defined($revid) && $revid eq "0") {
+		# The revid can be "0" for pages not in namespace 0 in which
+		# case another query is required to get the actual real revid
+		# Note that the pageid can also be "0" in this case so we must
+		# use titles and not pageids to do the second query
+		$query = {
+			action => 'query',
+			titles => $result->{query}->{recentchanges}[0]->{title},
+			prop => 'revisions',
+			rvlimit => '1',
+			rvdir => 'older',
+		};
+		my $result = $mediawiki->api($query);
+		my (undef, $v) = each %{$result->{query}->{pages}};
+		$revid = $v->{revisions}[0]->{revid};
+	}
+	return $revid;
 }
 
 # Get the last remote revision concerning the tracked pages and the tracked
---
